_posts/2018-02-05-chaining-http-requests-in-elm.md

```diff
@@ -0,0 +1,342 @@
+
+
+```haskell
+toHttpTask : Http.Request a -> Task String a
+toHttpTask request =
+    request
+        |> Http.toTask
+        |> Task.mapError toString
+
+fetchEvents : Task String (List String)
+fetchEvents =
+    toHttpTask eventsRequest
+
+fetchName : String -> Task String String
+fetchName login =
+    toHttpTask (nameRequest login)
+```
+
+`toHttpTask` is a common helper turning an `Http.Request` into a `Task`, transforming the `Http.Error` complex type into a serialized, purely textual version of it: a `String`.
+
+We'll also need a function allowing to extract the very first element of a list, if any, as we did in JavaScript using `events[0]`. Such a function is builtin the `List` core module as `List.head`. And let's make this function a `Task` too, as that will ease chaining things alltogether and allow us to expose an error message when the list is empty:
+
+```haskell
+pickFirst : List String -> Task String String
+pickFirst logins =
+    case List.head logins of
+        Just login ->
+            Task.succeed login
+
+        Nothing ->
+            Task.fail "No events."
+```
+
+Note the use of `Task.succeed` and `Task.fail`, which are approximately the Elm equivalents of `Promise.resolve` and `Promise.reject`.
+
+So in order to chain all the pieces we have so far, we obviously need *glue*. And this glue is the `Task.andThen` function. Its signature is:
+
+```haskell
+andThen : (a -> Task x b) -> Task x a -> Task x b
+```
+
+The first argument, the `(a -> Task x b)` function, will have to accept the raw result of a previous succesful task (the second argument), and return another task exposing another value. The whole thing is a new `Task` exposing that new value. That means we can finally chain our tasks this fancy way:  
+
+```haskell
+fetchEvents
+    |> Task.andThen pickFirst
+    |> Task.andThen fetchName
+```
+
+Neat. But wait. How are we going to execute this macro-task? Here comes the `Task.attempt` function, which signature is:
+
+```haskell
+attempt : (Result x a -> msg) -> Task x a -> Cmd msg
+```
+
+Woah, where does this `Result` come from? And what is this `Cmd msg` thing?
+
+- `Result x a` is the [Result] from the actual execution of our HTTP request, `x` being the error and `a` the succesfully received value,
+- `msg` here is a message, something we don't have defined yet; it's gonna be a value from a `Msg` type,
+- `Cmd msg` is the way in Elm to [schedule execution of side effects](https://www.elm-tutorial.org/en/03-subs-cmds/02-commands.html).
+
+So let's define a `Msg` type:
+
+```haskell
+type Msg
+    = Name (Result String String)
+```
+
+Don't forget types are functions, and `Name` conforms to the `(Result x a -> msg)` type. So we can use it with `Task.attempt`:
+
+```haskell
+fetchEvents
```

Comment:
Maybe adding a paragraph below stating what we are doing? Something like: "ok so, we start by picking the first even, then if the Task succeeds, we fetch the name, then we…"